_posts/2025-08-16-accessible-error-messages-in-forms.md

```diff
@@ -0,0 +1,158 @@
+
+The blog post [*Blind people don't visit my website*](https://www.a11y-collective.com/blog/blind-people-dont-visit-my-website/) lists frustrations from web users, such as:
+*"Error messages that come afterwards and eventually don't explain what you need to change."*
+
+Don’t make users guess — offer clear help in plain language, at the right moment. Tell users in advance which documents they'll need, and which fields are or aren't required. Add descriptions to fields where necessary.
+
+## How to show which fields are required?
+
+If you only ask what's really needed, all fields could technically be required — but that’s not always the case.
+
+### What’s the best way to indicate required?
+
+Many websites use indicators like “required”, “optional”, or an asterisk (\*). But which is clearest? 
+
+Do user research for your language. For example in Dutch the word "optioneel" (optional) is hard to understand for people with low literacy, and "niet verplicht" (not required) is a better choice. This may be different for your language.
+
+Prefer **“required”** over an asterisk, because an asterisk assumes prior knowledge of the meaning of that icon. If you do use an asterisk, explain its meaning above the form.
+
+### Mark required or non-required fields?
+
+Which is better — indicate required or non-required fields? It depends. Your choice should be based on the form’s function, your CMS or plugin, and user research.
+
+An option could be: when most fields are required, mark the *optional* ones. If most fields are not required, mark the *required* ones.
+
+{: .example }
+For example: "Please complete all fields. If a field is not required, it will be labeled "(Optional)".
+
+Whatever you choose, be consistent within the form and across all forms on your website. Inform users above the form how field requirements are indicated.
+
+{: .note }
+Please note: for **screen readers** is necessary to always add the attributes [`aria-required`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-required) and [`aria-invalid`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-invalid) for required form field.
+
+## When should I check for errors?
+
+Websites can check for errors while typing, after the focus moves out of a field, or after the form is submitted. Which approach is most user-friendly?
+
+Checking during typing can be confusing. Imagine typing an email address and seeing “Invalid email address” after the first letter. Of course, it’s invalid — you haven’t finished yet! 
+
+That leaves two options: check after leaving a field, or after submission. These can be combined. For certain fields — like dates that must be in the future — immediate checking is helpful. But checking after submission is always necessary.
```

Comment:
Worth noting that certain types of checks can happen while typing, e.g. length minimum/maximums. But when they do, care needs to be taken in how the feedback is reported.